### ISSUE TITLE
Strange exception behaviour

### ISSUE THREAD

**Issue by hreuver** [enasequence/cramtools#16]
Hi all,

We have accumulated quite a lot of BAM files over the years and we would like (need) to archive the old(er) ones. We have tested CRAMtools 2.1 and we are very happy with the results. Since we are talking about thousands of BAM files, we cannot go without automation. But...

The problem is as followed:
- when I run CRAMtools in a job file, it terminates immediatly (throwing an exception).
- when I copy the command from the job file and execute it on the head node or on the node it previously failed on, I have no issues what so ever.

My job file looks like this:
#PBS -N DNAXXXXXX
#PBS -j oe
#PBS -o /data/results/CRAM_archive/2013/october//DNAXXXXXX.log
#PBS -r y
#PBS -q bs_secondary
#PBS -m ea
#PBS -M XXXXXXXXXX@radboudumc.nl
#PBS -l nodes=1:ppn=12
cp /data/results/BGI/DNAXXXXXX/DNAXXXXXX.bam /scratch/
cp /data/results/BGI/DNAXXXXXX/DNAXXXXXX.bam.bai /scratch/
java -jar /share/apps/cramtools-2.1.jar cram -I /data/results/BGI/DNAXXXXXX/DNAXXXXXX.bam -O /scratch/DNAXXXXXX.cram -R /data/references/GrCh37/GrCh37_reference.fa --preserve-read-names --capture-all-tags -L *8
java -cp /share/apps/cramtools-2.1.jar net.sf.cram.ValidateCramFile -I /scratch/DNAXXXXXX.cram -R /data/references/GrCh37/GrCh37_reference.fa
mv /scratch/DNAXXXXXX.cram /data/results/CRAM_archive/2013/october//
rm -f /scratch/DNAXXXXXX.bam
rm -f /scratch/DNAXXXXXX.bam.bai

The exception I get when submitting the job file to torque:
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at net.sf.cram.CramTools.invoke(CramTools.java:93)
        at net.sf.cram.CramTools.main(CramTools.java:123)
Caused by: java.lang.RuntimeException: Uknown read or base category: 6
        at net.sf.cram.lossy.QualityScorePreservation.parseSinglePolicy(QualityScorePreservation.java:155)
        at net.sf.cram.lossy.QualityScorePreservation.parsePolicies(QualityScorePreservation.java:79)
        at net.sf.cram.lossy.QualityScorePreservation.<init>(QualityScorePreservation.java:39)
        at net.sf.cram.Bam2Cram.main(Bam2Cram.java:273)
        ... 6 more

I have tried to include the full path to java to make sure the same version is used in all cases, but the problem remains. The issue only seem to occur when applied to Illumina (mapped with BWA) BAM files. When I apply the same job file on our SOLiD data (mapped with Lifescope) I have no issues.

Does anyone have an idea what could cause this problem and how I can fix it?

Many thanks,

Rick

**Comment by vadimzalunin** [comment]
Hi Rick,

The first thing which comes to my mind is that the lossy model command 
line options are chewed by shell somehow. Perhaps you could try and put 
single quotes around the model, like this: -L '*8'.
Let me know if this does not help.

Vadim

**Comment by hreuver** [comment]
Hi Vadim,

That was one of the first things that came to my mind as well, but since we were able to run job files generated by the same script on the same server, but on BAM files from a different mapper, I didn't expect this to solve the problem. I will give it a try next week, since our production server is queued up for the weekend.

I did copy the data of one sample to one of our other servers to give it a go, but for some reason I was unable to reproduce the problem. It's an exact copy of the job file and the data, so I really don't understand why it does work on that machine. I will have a closer look at it next week and report back the results.

Have a good weekend,

Rick

**Comment by vadimzalunin** [comment]
Could it be that linux environment is different, for example shell?
Another idea is that the file name expansion depends on nullglob env 
variable, and for example if no file name match *8 in the working 
directory then it will preserve *8 but otherwise replace it with the 
files matching the pattern.
I think using * in command line options was a bad idea, there should be 
another way.

Vadim

**Comment by hreuver** [comment]
Hi Vadim,

There don't seem to be any differences in the Linux environments between the cluster we do and don't have this issue. I did do some additional tests, one where I run the job file with -L *8 and one where I run with -L '*8' and the first one fails, the second one runs without issues.

It still doesn't explain why it did work before with other data though. Maybe I overlooked something in the environment or something, but it seems system depended.

But my problem is solved and I can continu using CRAM. Many thanks for your help!
